processor/src/test/java/org/mapstruct/ap/test/bugs/_2840/Issue2840Test.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._2840;
+
+import org.mapstruct.ap.testutil.IssueKey;
+import org.mapstruct.ap.testutil.ProcessorTest;
+import org.mapstruct.ap.testutil.WithClasses;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Filip Hrisafov
+ */
+@IssueKey("2840")
+@WithClasses({
+    Issue2840Mapper.class,
+})
+class Issue2840Test {
+
+    @ProcessorTest
+    void shouldUseMethodWithMostSpecificReturnType() {
+        Issue2840Mapper.Target target = Issue2840Mapper.INSTANCE.map( (short) 10, 50 );
```

Comment:
Should we create an update method as well? Does that make sense at all (or is that cover by the `InheritanceSelector` ? Man.. I really need to get back into this 😄

Reply:
Adding an update method isn't really needed. We are testing the picking of the custom conversion functions. Which means that the return type is always in those, it has nothing to do with the mapping method. The `InheritanceSelector` is for other things, like picking the right object factory method.